Fixes to deflate_quick algorithm when compiling with USE_MMAP

nmoinvaz opened the pull request:
This fixes the bug reported in #654 with deflate_quick when compiling with `USE_MMAP`

* Fixed avail_out == 0 conditional not returning `need_more` in deflate_quick.
* Fixed ending block when returning `need_more` caused problems with inflate.

So instead of ending the block each time the function returns, we check upon start to see if it is Z_FINISH, and if it is, and the last block not yet started it will end the previous  block and start the last block.

I have tested this against pigz-bench-python and with USE_MMAP and the x-ray file from silesia corpus.

nmoinvaz committed:
Fixed avail_out == 0 conditional not returning need_more in deflate_quick.
Fixed ending block when returning need_more caused problems with inflate.

So instead of ending the block each time the function returns to finish the last block, we check upon start to see if it is the last block and if the last block has been started, and if not it will close the previous block and start the last block.

Dead2 commented:
```
Activated multiple generated file mode. Source: silesia-small.tar
Level 1: /tmp/deflatebench-1.tmp 270.1 MiB / 283,253,760 B

 Runs: 20
 Levels: 1
 Trimworst: 10
```

```
# Baseline 738a6e5b07c0495b8652c77b4879145c083ea49f
   text    data     bss     dec     hex filename
  93246    1304      32   94582   17176 libz-ng.so.1

 Level   Comp   Comptime min/avg/max  Decomptime min/avg/max  Compressed size
 1      57.881%    1.754/1.779/1.790       1.082/1.098/1.106  163950540

 Level   Comp   Comptime min/avg/max  Decomptime min/avg/max  Compressed size
 1      57.881%    1.772/1.781/1.790       1.077/1.097/1.118  163950540
```

```
## PR/657 bb5dba370012cbf1c322c3e8cfb5eda2671dc2ee
   text    data     bss     dec     hex filename
  93454    1304      32   94790   17246 libz-ng.so.1

 Level   Comp   Comptime min/avg/max  Decomptime min/avg/max  Compressed size
 1      57.870%    1.770/1.793/1.810       1.079/1.100/1.114  163918126
```

This makes a big change to compressed output as well, the output is now 32414 bytes smaller.
Are those changes well enough understood?

nmoinvaz committed:
Don't write end of last block when returning finish_started.

nmoinvaz commented:
Yes, I think the changes only relate to when the start and end blocks are written. The initial problem with `deflate_quick` was partially related to the last block not being written properly in some cases. So I had changed it to write start and end blocks each time it entered and exit the function. This probably caused it to write more frequently than it should have. However, that had some problems as inflate did not like seeing an end block in certain instances when the function does not return `block_done` or `finish_done`. So instead, it is now changed that at the top of the loop it will write the end of previous block if the function is processing the last block and start of block hasn't been written. 

And in my last commit I thought about a potential problem, where an end block would be written during last block processing if it returns `finish_started`. I have tested it against both tests which previously produced problems for `deflate_quick`.
